homeassistant/components/shelly/event.py

```diff
@@ -95,42 +101,66 @@ async def async_setup_entry(
         if TYPE_CHECKING:
             assert coordinator
 
-        key_instances = get_rpc_key_instances(coordinator.device.status, RPC_EVENT.key)
+        if config_entry.data[CONF_SLEEP_PERIOD] and not coordinator.device.initialized:
+            ent_reg = er.async_get(hass)
+            entries = er.async_entries_for_config_entry(ent_reg, config_entry.entry_id)
+            for entry in entries:
+                if entry.domain != "event":
+                    continue
 
-        for key in key_instances:
-            if RPC_EVENT.removal_condition and RPC_EVENT.removal_condition(
-                coordinator.device.config, coordinator.device.status, key
-            ):
-                unique_id = f"{coordinator.mac}-{key}"
-                async_remove_shelly_entity(hass, EVENT_DOMAIN, unique_id)
-            else:
-                entities.append(ShellyRpcEvent(coordinator, key, RPC_EVENT))
+                # Only smoke sensors can have a test event entity
+                attribute = entry.unique_id.split("-")[-1]
+                if not attribute.startswith("smoke"):
+                    continue
 
-        script_instances = get_rpc_key_instances(
-            coordinator.device.status, SCRIPT_EVENT.key
-        )
-        for script in script_instances:
-            script_name = get_rpc_entity_name(coordinator.device, script)
-            if script_name == BLE_SCRIPT_NAME:
-                continue
+                entities.append(
+                    ShellyRpcTestEvent(coordinator, attribute, entry.original_name)
```

Comment:
We don't need the name when `has_entity_name` is `True`.
```suggestion
                    ShellyRpcTestEvent(coordinator, attribute)
```

Reply:
I tested this with a real device and I didn't see any problem. Can you describe what didn't work? Maybe I'm missing something.